core/aggregate/event_sourced_repository.go

```diff
@@ -0,0 +1,83 @@
+package aggregate
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/get-eventually/go-eventually/core/event"
+	"github.com/get-eventually/go-eventually/core/version"
+)
+
+// Factory is a function that creates new zero-valued
+// instances of an aggregate.Root implementation.
+type Factory[I ID, T Root[I]] func() T
+
+// EventSourcedRepository provides an aggregate.Repository interface implementation
+// that uses an event.Store to store and load the state of the Aggregate Root.
+type EventSourcedRepository[I ID, T Root[I]] struct {
+	eventStore event.Store
+	factory    Factory[I, T]
+}
+
+func NewEventSourcedRepository[I ID, T Root[I]](
+	eventStore event.Store,
+	factory Factory[I, T],
+) EventSourcedRepository[I, T] {
+	return EventSourcedRepository[I, T]{
+		eventStore: eventStore,
+		factory:    factory,
+	}
+}
+
+func (repo EventSourcedRepository[I, T]) Get(ctx context.Context, id I) (T, error) {
+	var zeroValue T
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	streamID := event.StreamID(id.String())
+	eventStream := make(chan event.Persisted, 1)
+
+	group, ctx := errgroup.WithContext(ctx)
+	group.Go(func() error {
+		if err := repo.eventStore.Stream(ctx, eventStream, streamID, version.SelectFromBeginning); err != nil {
+			return fmt.Errorf("%T: failed while reading event from stream: %w", repo, err)
+		}
+
+		return nil
+	})
+
+	root := repo.factory()
+
+	if err := root.rehydrate(root, eventStream); err != nil {
```

Comment:
>Why is it required to pass root additionally to using it as a receiver? 

Yeah, the reason for it was due to the lack of "abstract", let's say: `rehydrate` is a private method that only `BaseRoot` should implement, but `BaseRoot` would have no access to the `Aggregate`'s `Apply` method. So the `Root` should be passed explicitly.

Well, I've now resolved it by refactoring `rehydrate` in a package-level function, `RehydrateFromEvents`. This uses the same approach `RecordThat` uses.

Wdyt?

Reply:
Looks better now 👍🏻 
Btw, should `RehydrateFromEvents` be public if it's package-level?

Reply:
Yeah good point, I thought it could be useful to have it exported for extension modules outside `core`, but likely it's only required by `aggregate.EventSourcedRepository` 👍🏻 

Changed it in the last commits.